File: Chain of Responsibility/chain.rb
class Transaction
  attr_reader :amount, :currency

  def initialize(amount, currency)
    @amount = amount
    @currency = currency
  end
end

class BaseHandler
  attr_reader :successor

  def initialize(successor = nil)
    @successor = successor
  end

  def call(transaction)
    return successor.call(transaction) unless can_handle?(transaction)

    handle(transaction)
  end

  def handle(_transaction)
    raise NotImplementedError, 'Each handler should respond to handle and can_handle? methods'
  end
end

class GatewayA < BaseHandler

  private

  def handle(transaction)
    puts "handling the transaction with GatewayA"
  end


  def can_handle?(transaction)
    transaction.amount < 100 && transaction.currency == 'USD'
  end
end

class GatewayB < BaseHandler

  private

  def handle(transaction)
    puts "handling the transaction with GatewayB"
  end

  def can_handle?(transaction)
    transaction.amount >= 100
  end
end

transaction = Transaction.new(50, 'USD')

chain = GatewayA.new(GatewayB.new)
chain.call(transaction)

transaction = Transaction.new(200, 'USD');
chain.call(transaction);

File: Chain of Responsibility/chain.js
class Transaction {
  constructor(amount, currency) {
    this.amount = amount;
    this.currency = currency;
  }
}

class BaseHandler {
  constructor(next = null) {
    this.next = next;
  }

  call(transaction) {
    if (this.canHandle(transaction)) {
      return this.handle(transaction)
    }

    if (this.next) {
      return this.next.call(transaction);
    }
    return 'Failed';
  }
}

class GatewayA extends BaseHandler {
  constructor(next) {
    super(next);
  }

  handle(transaction) {
    return 'handling on GatewayA ';
  }

  canHandle(transaction) {
    return transaction.amount < 100;
  }
}

class GatewayB extends BaseHandler {
  constructor(next) {
    super(next);
  }

  handle(transaction) {
    return 'handling on GatewayB ';
  }

  canHandle(transaction) {
    return transaction.amount < 200;
  }
}

let transaction = new Transaction(100, 'INR');
chain = new GatewayA(new GatewayB());
console.log(
  chain.call(transaction)
);

